### PR TITLE
IO with filtered particles

### PR DESCRIPTION
**Pull request by psychocoderHPC**
- refactor HDF, ADIOS to allow usage as multi plugin
- refactor interface class and helper for multi plugins
  - `IHelp`: is the new factory to create a `ISlave`
  - `ISlave`: remove factory method
  - `Master`: use `IHelp` fectory method to create a instance of `ISlave`
- `IIOBackend`:
  - inherit from `ISlave`
  - `expandHelp()` is moved into `IHelp`
- refactor `EnergyParticle` to use new interafces
- add documentation for the new option
- add function to split a string ￼…
  - add `splitString` function to `plugins/misc`
  - add function `containsObject()` to check if a container holds an object
  - add function `removeSpaces()` to remove all spaces within a string
  - add collection include file `misc.hpp`

# Review

Please review the PR commit wise!

# Todo
- [x] add Doxygen documentation
- [x] update rst files
- [x] restart tests
- [x] io tests

### CODE REVIEW

File: include/picongpu/plugins/hdf5/HDF5Writer.hpp [hunk 1]
```diff
@@ -88,60 +94,284 @@ using namespace splash;
...
+        std::string concatenatedFilterNames;
+        std::vector< std::string > allowedFilters;
+
+        template<typename T_TupleVector>
+        struct CreateSpeciesFilter
+        {
+            using type = plugins::misc::SpeciesFilter<
+                typename pmacc::math::CT::At<
+                    T_TupleVector,
+                    bmpl::int_<0>
```
Comment (ax3l) [inline include/picongpu/plugins/hdf5/HDF5Writer.hpp]: we could re-use / refactor the enums and classes from #2368 here if we move them to a general PIConGPU place

File: include/picongpu/plugins/multi/HelpValue.hpp [hunk 2]
```diff
@@ -0,0 +1,154 @@
...
+        std::string getDefaultAsStr()
+        {
+            std::stringstream ss;
+            ss << getDefault( );
+            return ss.str();
+        }
+
+        /** get the value set by the user
+         *
+         * Throw an exption if there is no default value defined and idx is
```
Comment (ax3l) [inline include/picongpu/plugins/multi/HelpValue.hpp]: exception

File: include/picongpu/plugins/hdf5/HDF5Writer.hpp (hunk 1 above)
Comment (psychocoderHPC) [reply include/picongpu/plugins/hdf5/HDF5Writer.hpp]: we can address this later (future PR). all in all it is something like an apply maybe we can write it more boost standard conform.

File: include/picongpu/plugins/EnergyParticles.hpp [hunk 3]
```diff
@@ -335,18 +361,26 @@ namespace picongpu
...
                 );
 
-                desc.add_options( )(
-                    ( masterPrefix + prefix + ".period").c_str( ),
-                    boost::program_options::value< std::vector< uint32_t > >( &notifyPeriod )->multitoken( ),
-                    "compute kinetic and total energy [for each n-th step] enable plugin by setting a non-zero value"
+                notifyPeriod.registerHelp(
+                    desc,
+                    masterPrefix + prefix,
+                    std::string( "[" ) + concatenatedFilterNames + "]"
```
Comment (psychocoderHPC) [inline include/picongpu/plugins/EnergyParticles.hpp]: BUG: notify should not have a extended description

File: include/picongpu/plugins/EnergyParticles.hpp [hunk 4]
```diff
@@ -335,18 +361,26 @@ namespace picongpu
...
                 );
-                desc.add_options( )(
-                    ( masterPrefix + prefix + ".filter").c_str( ),
-                    boost::program_options::value< std::vector< std::string > >( &filter )->multitoken( ),
-                    ( std::string( "particle filter: " ) + concatenatedFilterNames ).c_str()
+                filter.registerHelp(
+                    desc,
+                    masterPrefix + prefix,
+                    concatenatedFilterNames
```
Comment (psychocoderHPC) [inline include/picongpu/plugins/EnergyParticles.hpp]: must be `std::string( "[" ) + concatenatedFilterNames + "]"`

File: docs/source/usage/plugins/adios.rst [hunk 5]
```diff
@@ -28,19 +28,30 @@ One can e.g. disable the output of particles by setting:
...
 ``--adios.ost``              Set number of I/O OSTs for ADIOS ``MPI_AGGREGATE`` transport method.
+``--adios.disable-meta``     Disable on-defly creation of the adios journal file.
+``--adios.source``           Select data sources to dump. Default is ``species_all,fields_all``, which dumps all fields and particle species.
 ============================ ==================================================================================================================================================================
 
+.. note::
+
+   This plugin is a multi plugin. 
+   Command line parameter can be used multiple times to create e.g. dumps with different dumping period.
+   In the case where a optional parameter with a default value is explicitly defined the parameter will be always passed to the instance of the multi plugin where the parameter is not set.
```
Comment (ax3l) [inline docs/source/usage/plugins/adios.rst]: Maybe replace sentence with:

```
Attention when defining optional parameters!
When an optional parameter is added, is value is always assigned to the *leftmost* plugin inside of that multi-plugin.
As an example, the command line: ``...`` creates two plugins:

#. dump all ...
#. dump all ...
```

File: docs/source/usage/plugins/hdf5.rst [hunk 6]
```diff
@@ -44,18 +44,26 @@ One can e.g. disable the output of particles by setting:
...
 PIConGPU command line option Description
-============================ ======================================================================================
+============================ ================================================================================================================
 ``--hdf5.period``            Period after which simulation data should be stored on disk.
-                             Default is ``0``, which means that no data is stored.
 ``--hdf5.file``              Relative or absolute fileset prefix for simulation data.
                              If relative, files are stored under ``simOutput/``.
-                             Default is ``h5``.
-============================ ======================================================================================
+``--hdf5.source``            Select data sources to dump. Default is ``species_all,fields_all``, which dumps all fields and particle species.
```
Comment (ax3l) [inline docs/source/usage/plugins/hdf5.rst]: feel free to continue on the next line (in the 2nd column) after the comma in a table

File: docs/source/usage/plugins/hdf5.rst [hunk 7]
```diff
@@ -44,18 +44,26 @@ One can e.g. disable the output of particles by setting:
...
+============================ ================================================================================================================
+
+.. note::
+
+   This plugin is a multi plugin. 
+   Command line parameter can be used multiple times to create e.g. dumps with different dumping period.
+   In the case where a optional parameter with a default value is explicitly defined the parameter will be always passed to the instance of the multi plugin where the parameter is not set.
```
Comment (ax3l) [inline docs/source/usage/plugins/hdf5.rst]: see above

File: include/picongpu/plugins/adios/ADIOSWriter.hpp [hunk 8]
```diff
@@ -144,6 +149,248 @@ int64_t defineAdiosVar(int64_t group_id,
...
+            return std::shared_ptr< ISlave >(
+                new ADIOSWriter(
+                    help,
+                    id,
+                    cellDescription
+                )
+            );
+        }
+
+        //! periodicity of computing the particle energy
```
Comment (ax3l) [inline include/picongpu/plugins/adios/ADIOSWriter.hpp]: typo: this is the ADIOS and not the energy plugin :)

File: include/picongpu/plugins/adios/ADIOSWriter.hpp [hunk 9]
```diff
@@ -144,6 +149,248 @@ int64_t defineAdiosVar(int64_t group_id,
...
+            ""
+        };
+
+        plugins::multi::Option< std::string > compression = {
+            "compression",
+            "ADIOS compression method, e.g., zlib (see `adios_config -m` for help)",
+            "none"
+        };
+
+        bool isIndependent = false;
```
Comment (ax3l) [inline include/picongpu/plugins/adios/ADIOSWriter.hpp]: please add a doxygen comment on this var what `isIndependent` stands for

File: include/picongpu/plugins/hdf5/HDF5Writer.hpp [hunk 10]
```diff
@@ -88,60 +94,265 @@ using namespace splash;
...
+            return std::shared_ptr< ISlave >(
+                new HDF5Writer(
+                    help,
+                    id,
+                    cellDescription
+                )
+            );
+        }
+
+        //! periodicity of computing the particle energy
```
Comment (ax3l) [inline include/picongpu/plugins/hdf5/HDF5Writer.hpp]: typo: this is the HDF5 plugin

File: include/picongpu/plugins/hdf5/HDF5Writer.hpp [hunk 11]
```diff
@@ -88,60 +94,265 @@ using namespace splash;
...
+            "data sources: ",
+            "species_all, fields_all"
+        };
+
+        plugins::multi::Option< std::string > fileName = {
+            "file",
+            "HDF5 output filename (prefix)"
+        };
+
+        bool isIndependent = false;
```
Comment (ax3l) [inline include/picongpu/plugins/hdf5/HDF5Writer.hpp]: pls also add a doxygen doc string here

File: docs/TBG_macros.cfg [hunk 12]
```diff
@@ -166,23 +166,26 @@ TBG_macroCount="--<species>_macroParticlesCount.period 100"
...
+TBG_adios="--adios.period 100 --adios.file simData --adios.source 'species_all,fields_all'"
 # see 'adios_config -m', e.g., for on-the-fly zlib compression
 #     (compile ADIOS with --with-zlib=<ZLIB_ROOT>)
 #   --adios.compression zlib
 # for parallel large-scale parallel file-systems:
 #   --adios.aggregators <N * 3> --adios.ost <N>
 # avoid writing meta file on massively parallel runs
-#   --adios.disable-meta
+#   --adios.disable-meta <B>
+#   B = 0 is equal to false, B = 1 is true
```
[the remark this answers is not in the document]
Comment (psychocoderHPC) [reply docs/TBG_macros.cfg]: no not possible anymore :-(

File: docs/source/usage/plugins/adios.rst (hunk 5 above)
Comment (psychocoderHPC) [reply docs/source/usage/plugins/adios.rst]: fixed

File: docs/source/usage/plugins/hdf5.rst (hunk 7 above)
Comment (psychocoderHPC) [reply docs/source/usage/plugins/hdf5.rst]: fixed

File: include/picongpu/plugins/misc/removeSpaces.hpp [hunk 13]
```diff
@@ -0,0 +1,51 @@
...
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with PIConGPU.
+ * If not, see <http://www.gnu.org/licenses/>.
+ */
+
+#pragma once
+
+#include <string>
+
```
[the remark this answers is not in the document]
Comment (psychocoderHPC) [reply include/picongpu/plugins/misc/removeSpaces.hpp]: added <algorithm>

File: include/picongpu/plugins/misc/splitString.hpp [hunk 14]
```diff
@@ -0,0 +1,64 @@
...
+     *
+     * Based on Stack Overflow post:
+     *   source: https://stackoverflow.com/a/28142357
+     *   author: Marcin
+     *   date: Jan 25 '15
+     *
+     * @param input string to split
+     * @param regex separator between two elements
+     */
+    std::vector< std::string > splitString(
```
[the remark this answers is not in the document]
Comment (psychocoderHPC) [reply include/picongpu/plugins/misc/splitString.hpp]: done

File: include/picongpu/plugins/misc/splitString.hpp [hunk 15]
```diff
@@ -0,0 +1,64 @@
...
+ * You should have received a copy of the GNU General Public License
+ * along with PIConGPU.
+ * If not, see <http://www.gnu.org/licenses/>.
+ */
+
+#pragma once
+
+#include <string>
+#include <regex>
+
```
[the remark this answers is not in the document]
Comment (psychocoderHPC) [reply include/picongpu/plugins/misc/splitString.hpp]: done

File: include/picongpu/plugins/hdf5/HDF5Writer.hpp [hunk 16]
```diff
@@ -290,7 +294,7 @@ class HDF5Writer :
 
     /** constructor
      *
-     * @param isIndependent if `true`: class register itself to PluginConnector
+     * @param selfRegister if `true`: class register itself to PluginConnector
```
Comment (ax3l) [inline include/picongpu/plugins/hdf5/HDF5Writer.hpp]: register**s**

File: include/picongpu/plugins/hdf5/HDF5Writer.hpp [hunk 17]
```diff
@@ -138,7 +137,12 @@ class HDF5Writer :
             "HDF5 output filename (prefix)"
         };
 
-        bool isIndependent = false;
+        /** defines if the plugin must be register it self to the PMacc plugin system
```
Comment (ax3l) [inline include/picongpu/plugins/hdf5/HDF5Writer.hpp]: ... must register itself ...

File: include/picongpu/plugins/hdf5/HDF5Writer.hpp [hunk 18]
```diff
@@ -138,7 +137,12 @@ class HDF5Writer :
...
+         *
+         * true = the plugin is registering it self
```
Comment (ax3l) [inline include/picongpu/plugins/hdf5/HDF5Writer.hpp]: itself

File: include/picongpu/plugins/hdf5/HDF5Writer.hpp [hunk 19]
```diff
@@ -138,7 +137,12 @@ class HDF5Writer :
...
+         * false = the plugin is not registering it self (plugin is controlled by another class)
```
Comment (ax3l) [inline include/picongpu/plugins/hdf5/HDF5Writer.hpp]: itself

File: include/picongpu/plugins/hdf5/HDF5Writer.hpp (hunk 18 above)
Comment (ax3l) [reply include/picongpu/plugins/hdf5/HDF5Writer.hpp]: oh oh, forgotten ;)

**Pull request merged**